### PR TITLE
[SPARK-6129][MLLIB][DOCS] Added user guide for evaluation metrics

### PR DESCRIPTION
**Pull request by sethah**

### CODE REVIEW

File: docs/mllib-guide.md [hunk 1]
```diff
@@ -48,6 +48,7 @@ This lists functionality included in `spark.mllib`, the main MLlib API.
 * [Feature extraction and transformation](mllib-feature-extraction.html)
 * [Frequent pattern mining](mllib-frequent-pattern-mining.html)
   * FP-growth
+* [Metrics](mllib-metrics.html)
```
Comment (srowen) [inline docs/mllib-guide.md]: Evaluation Metrics? Metrics makes me thing of app runtime performance metrics

File: docs/mllib-metrics.md [hunk 2]
```diff
@@ -0,0 +1,1464 @@
...
+---
+
+* Table of contents
+{:toc}
+
+
+## Algorithm Metrics
+
+Spark's MLlib comes with a number of machine learning algorithms that can be used to learn from and make predictions
+on data. When applying these algorithms, there is a need to evaluate their performance on certain criteria, depending
```
Comment (srowen) [inline docs/mllib-metrics.md]: Nit, but maybe worth saying we're evaluating the output -- the model -- not the algorithm itself. Evaluations also aren't a property of the application that consumes the model, but of the model.

File: docs/mllib-metrics.md [hunk 3]
```diff
@@ -0,0 +1,1464 @@
...
+dataset into one of two possible groups (e.g. fraud or not fraud) and is a special case of multiclass classification.
+Most binary classification metrics can be generalized to multiclass classification metrics.
+
+<table class="table">
+  <thead>
+    <tr><th>Metric</th><th>Definition</th></tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td>Precision (Postive Predictive Value)</td>
```
Comment (srowen) [inline docs/mllib-metrics.md]: Terms like TP aren't explained and won't be obvious to the reader that turns to this document, I think, for an explanation. The mathematical definition of things like ROC is OK but this offers no intuition about them. I don't think we need to reproduce a text on what they mean, but at least a hyperlink to wikipedia?

File: docs/mllib-metrics.md [hunk 4]
```diff
@@ -0,0 +1,1464 @@
...
+<div class="codetabs">
+The following code snippets illustrate how to evaluate the performance of a multilabel classifer.
+
+<div data-lang="scala" markdown="1">
+
+{% highlight scala %}
+import org.apache.spark.mllib.evaluation.MultilabelMetrics
+import org.apache.spark.rdd.RDD;
+
+/**
```
Comment (srowen) [inline docs/mllib-metrics.md]: This might be a nit, but javadoc inside example code on a web page doesn't seem necessary. This setup applies to all 3 languages right? it can be lifted out of the code and explained in prose above?

**Comment by srowen** [comment]
I really like a good write up on evaluation metrics. My broad suggestions are to consider less math and more intuitive explanation, without getting into a whole lot of writing. People can read a text or wikipedia for the math; that's not quite the MLlib docs audience. Also I think you can refactor out some of the duplicated comments.

**Comment by sethah** [comment]
Sean,

Thanks for your feedback. I agree that more intuitive descriptions will be helpful, and so I will work on getting those into the document. One thing I'd like to point out is that I think having clean mathematical formulas for each algorithm is necessary because some of these algorithms are implemented differently in MLlib than they are defined on, say Wikipedia. I think having a description, or as you said, some links alongside the mathematical definitions is best. If you disagree, let me know, otherwise I'll move forward with that thought process in mind.

**Comment by srowen** [comment]
Yes, that sounds important where the definition or math may not be standard or obvious. I think most people reading the doc won't have that background, so a few words or a link to the idea behind each metric can help, yes.

File: docs/mllib-guide.md (hunk 1 above)
Comment (sethah) [inline docs/mllib-guide.md]: Changed to evaluation metrics.

File: docs/mllib-metrics.md (hunk 2 above)
Comment (sethah) [inline docs/mllib-metrics.md]: Corrected this to say evaluation of the model.

File: docs/mllib-metrics.md (hunk 4 above)
Comment (sethah) [inline docs/mllib-metrics.md]: Moved duplicated comments to the markdown text.

**Comment by sethah** [comment]
Sean,

I added a bit of background on things like TP, FP, precision, recall, ROC, etc... to the guide. I tried to explain the base concepts for classification since the different flavors of classification algo metrics basically just extend the basic ideas of precision, recall, etc. I also added an explanation of each of the ranking metric algorithms since those are not as well defined/easy to find on the internet. Additionally, I added hyperlinks to further reading on these topics via wikipedia.

I left all the math definitions; I wasn't clear if you were suggesting that we only leave some of them in or just supplement them with explanations. I didn't think it made a ton of sense to define some of them but not all. Also, I find it useful to see mathematical representations of what the algorithms take as parameters. Let me know if you'd like to remove more of the math (the notation can get a bit heavy) or if you think it's too wordy.

Thanks!

File: docs/mllib-evaluation-metrics.md [hunk 5]
```diff
@@ -0,0 +1,1475 @@
...
+
+### Binary Classification
+
+[Binary classifiers](https://en.wikipedia.org/wiki/Binary_classification) are used to separate the elements of a given
+dataset into one of two possible groups (e.g. fraud or not fraud) and is a special case of multiclass classification.
+Most binary classification metrics can be generalized to multiclass classification metrics.
+
+#### Threshold Tuning
+
+It is important to understand that, in most classification models, the output of the model is actually a probability or
```
Comment (srowen) [inline docs/mllib-evaluation-metrics.md]: I think the text is fine as-is, but I would take a little issue with stating that the output is usually a probability. It isn't for several common models like the SVM or decision trees, but it also isn't necessary that it be a probability for the discussion here -- just a number that's higher when the model thinks the positive class is more likely. But I wouldn't edit it just for this

File: docs/mllib-evaluation-metrics.md [hunk 6]
```diff
@@ -0,0 +1,1475 @@
...
+
+A [multiclass classification](https://en.wikipedia.org/wiki/Multiclass_classification) describes a classification
+problem where there are $M \gt 2$ possible labels for each data point (the case where $M=2$ is the binary
+classification problem). For example, classifying handwriting samples to the digits 0 to 9, having 10 possible classes.
+
+#### Label based metrics
+
+Opposed to binary classification where there are only two possible labels, multiclass classification problems have many
+possible labels and so the concept of label-based metrics is introduced. Overall precision measures precision across all
+labels -  the number of times any class was predicted correctly (true positives) normalized by the number of data
```
Comment (srowen) [inline docs/mllib-evaluation-metrics.md]: Is this really called "precision" and not just "accuracy"? fine if it is, it was just unfamiliar to me. I understand one-label precision, yes.

File: docs/mllib-evaluation-metrics.md [hunk 7]
```diff
@@ -0,0 +1,1475 @@
...
+{% endhighlight %}
+
+</div>
+</div>
+
+## Regression Model Evaluation
+
+[Regression analysis](https://en.wikipedia.org/wiki/Regression_analysis), most commonly formulated as a least-squares
+minimization, involves estimating the relationships among input and output variables. In this context, regression
+metrics refer to a continuous output variable.
```
Comment (srowen) [inline docs/mllib-evaluation-metrics.md]: This may just be my preference of wording, but lots of models involve estimating the relationship between input and output. The key point is a little obscured here, that regression is specifically the case where the output is continuous. Least-squares minimization isn't an essential property of regression. All of that is correct here, just felt like it could be more direct.

File: docs/mllib-evaluation-metrics.md [hunk 8]
```diff
@@ -0,0 +1,1475 @@
...
+      <td>Root Mean Squared Error (RMSE)</td>
+      <td>$RMSE = \sqrt{\frac{\sum_{i=0}^{N-1} (\mathbf{y}_i - \hat{\mathbf{y}}_i)^2}{N}}$</td>
+    </tr>
+    <tr>
+      <td>Coefficient of Determination $(R^2)$</td>
+      <td>$R^2=1 - \frac{MSE}{\text{VAR}(\mathbf{y}) \cdot (N-1)}=1-\frac{\sum_{i=0}^{N-1}
+        (\mathbf{y}_i - \hat{\mathbf{y}}_i)^2}{\sum_{i=0}^{N-1}(\mathbf{y}_i-\bar{\mathbf{y}})^2}$</td>
+    </tr>
+    <tr>
+      <td>Mean Absoloute Error (MAE)</td>
```
Comment (srowen) [inline docs/mllib-evaluation-metrics.md]: Nit: Move MAE up with RMSE? they go together.

**Comment by srowen** [comment]
Looking good. I left minor comments which I don't feel strongly about. I did not try running the code examples. We'll need to double-check the docs build and look right on merge.

File: docs/mllib-evaluation-metrics.md (hunk 8 above)
Comment (sethah) [inline docs/mllib-evaluation-metrics.md]: Done.

File: docs/mllib-evaluation-metrics.md (hunk 7 above)
Comment (sethah) [inline docs/mllib-evaluation-metrics.md]: I've reworded to make it simpler.

File: docs/mllib-evaluation-metrics.md (hunk 6 above)
Comment (sethah) [inline docs/mllib-evaluation-metrics.md]: Accuracy and precision (and recall for that matter) end up being the same thing for multiclass classification. That's why the per-label versions are better. But at any rate, these methods are available in MLlib and so are defined here.

File: docs/mllib-evaluation-metrics.md (hunk 5 above)
Comment (sethah) [inline docs/mllib-evaluation-metrics.md]: I agree and I've reworded with your points in mind.

**Comment by srowen** [comment]
LGTM. Let's leave it for a bit to collect other comments.

File: docs/mllib-evaluation-metrics.md [hunk 9]
```diff
@@ -0,0 +1,1476 @@
...
+layout: global
+title: Evaluation Metrics - MLlib
+displayTitle: <a href="mllib-guide.html">MLlib</a> - Evaluation Metrics
+---
+
+* Table of contents
+{:toc}
+
+
+## Algorithm Metrics
```
Comment (mengxr) [inline docs/mllib-evaluation-metrics.md]: Maybe we can remove the section title. I was wondering how `Algorithm Metrics` differs from `Evaluation Metrics` when I first saw it.

File: docs/mllib-evaluation-metrics.md [hunk 10]
```diff
@@ -0,0 +1,1476 @@
...
+metrics that are currently available in Spark's MLlib are detailed in this section.
+
+## Classification Model Evaluation
+
+While there are many different types of classification algorithms, the evaluation of classification models all share
+similar principles. In a [supervised classification problem](https://en.wikipedia.org/wiki/Statistical_classification),
+there exists a true output and a model-generated predicted output for each data point. For this reason, the results for
+each data point can be assigned to one of four categories:
+
+* True Positive (TP) - class predicted by model and class in true output
```
Comment (mengxr) [inline docs/mllib-evaluation-metrics.md]: This is confusing to me: `class predicted by model and class in true output`. I think the main issue is the term `true output`, which is not used in MLlib doc before. The following is easier to understand (at least to me):

* TP: label is positive and prediction is also positive
* TN: label is negative and prediction is also negative
* FP: label is negative but prediction is positive
* FN: label is positive but prediction is negative

For multiclass problems, we can define what `positive` and `negative` means under the context of a particular class.

File: docs/mllib-evaluation-metrics.md [hunk 11]
```diff
@@ -0,0 +1,1476 @@
...
+import org.apache.spark.mllib.classification.LogisticRegressionWithLBFGS
+import org.apache.spark.mllib.evaluation.BinaryClassificationMetrics
+import org.apache.spark.mllib.regression.LabeledPoint
+import org.apache.spark.mllib.util.MLUtils
+
+// Load training data in LIBSVM format
+val data = MLUtils.loadLibSVMFile(sc, "data/mllib/sample_binary_classification_data.txt")
+
+// Split data into training (60%) and test (40%)
+val splits = data.randomSplit(Array(0.6, 0.4), seed = 11L)
```
Comment (mengxr) [inline docs/mllib-evaluation-metrics.md]: minor: In Scala, we can use `val Array(training, test) = data.randomSplit(...)` and then `training.cache()`. It is similar in Python.

File: docs/mllib-evaluation-metrics.md [hunk 12]
```diff
@@ -0,0 +1,1476 @@
...
+  val prediction = model.predict(features)
+  (prediction, label)
+}
+
+// Instantiate metrics object
+val metrics = new BinaryClassificationMetrics(predictionAndLabels)
+
+// Precision by threshold
+val precision = metrics.precisionByThreshold
+precision.foreach(x => printf("Threshold: %1.2f, Precision: %1.2f\n", x._1, x._2))
```
Comment (mengxr) [inline docs/mllib-evaluation-metrics.md]: Since this is user guide, we should keep the examples minimal. So I don't think it is necessary to format the numbers. For example, the following is easier to read:

~~~scala
precision.foreach { case (t, p) =>
  println(s"Threshold: $t, Precision: $p")
}
~~~

**Comment by mengxr** [comment]
@sethah This is great! I went over the generated doc. Just some minor comments:

1. Could you update the capitalization for section tiles? This is to be consistent with other MLlib guides. For example, we only write `Evaluation Metrics` in the display title, then `Evaluation metrics` in section titles and some other places.
2. Could you remove the number formatting? It is useful but it might distract readers. Also, for evaluation metrics, sometimes having 2 or 3 digits is not sufficient.

**Comment by mengxr** [comment]
Btw, you should consider splitting the PR into small ones next time. It is easier to review 4~5 small PRs than one big PR combining all of them:)

File: docs/mllib-evaluation-metrics.md (hunk 12 above)
Comment (sethah) [inline docs/mllib-evaluation-metrics.md]: I have updated all scala code to use string interpolation. The Java and Python code still use string formatting but I have not truncated the decimals like before.

File: docs/mllib-evaluation-metrics.md (hunk 11 above)
Comment (sethah) [inline docs/mllib-evaluation-metrics.md]: Fixed this in python and scala.

File: docs/mllib-evaluation-metrics.md (hunk 10 above)
Comment (sethah) [inline docs/mllib-evaluation-metrics.md]: I switched up the explanations to what you have above. I added some explanations to the multiclass section to define what a positive and negative label mean. May be worth having someone besides me take a look to make sure it's not confusing.

**Comment by sethah** [comment]
@mengxr I updated the guide with your suggestions. I corrected the capitalization scheme and other things you listed. Let me know if you find anything else.

Thanks for the tip on splitting PRs, will definitely keep that in mind for future PRs. Would have made this one a bit more manageable as well!

**Comment by mengxr** [comment]
Merged into master. Thanks @sethah for writing this up and @srowen for reviewing!

**Comment by mengxr** [comment]
@sethah Ah, just realized one minor issue. `print abc` doesn't work in Python 3. We have to use `print(abc)`. Could you create a new JIRA for this? Thanks!

**Comment by sethah** [comment]
@mengxr I have created the [JIRA](https://issues.apache.org/jira/browse/SPARK-9490). I may get a chance to fix that sometime next week.
